### ISSUE TITLE
[RDKit Error] unable to avoid 

### ISSUE THREAD

**Issue by ashyrmamatov01** [rdkit/rdkit#4308]
**Describe the bug**
I am working with a machine learning model which generates SMART-like patterns that can be converted to SMILES. Obviously, some of the predictions cannot be converted to SMILES due to incorrect predictions. 

The problem is that some of the RDKit Errors cannot be avoided and gives 
```*** Error in `python': double free or corruption (out): 0x00007fffe28774e0 ***```
Mainly, error comes when this function ```Chem.MolFromSmarts(smarts)```.

**To Reproduce**
```
import re
from rdkit import Chem

normal_smarts = "[CH3;!R;D1]-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH3;!R;D1] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[O;!R;D2] [O;!R;D2](-[C;!R;D3])-[CH2;!R;D2] [C;!R;D3](-[CH2;!R;D2])(=[O;!R;D1])-[O;!R;D2] ( = [O;!R;D1]=[C;!R;D3] ) [CH2;!R;D2](-[C;!R;D3])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[C;!R;D3])-[CH2;!R;D2] [C;!R;D3](-[CH2;!R;D2])(=[O;!R;D1])-[O;!R;D2] ( = [O;!R;D1]=[C;!R;D3] ) [O;!R;D2](-[C;!R;D3])-[CH2;!R;D2] [CH2;!R;D2](-[CH;!R;D3])-[O;!R;D2] [CH;!R;D3](-[CH2;!R;D2])(-[CH2;!R;D2])-[CH2;!R;D2] ( [CH2;!R;D2](-[CH2;!R;D2])-[CH;!R;D3] [CH2;!R;D2](-[CH2;!R;D2])-[CH3;!R;D1] [CH3;!R;D1]-[CH2;!R;D2] ) [CH2;!R;D2](-[CH2;!R;D2])-[CH;!R;D3] [CH2;!R;D2](-[CH2;!R;D2])-[CH3;!R;D1] [CH3;!R;D1]-[CH2;!R;D2]"

still_ok_smarts = "[CH3;!R;D1]-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH3;!R;D1] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[O;!R;D2] ( [CH3;!R;D1]-[CH;!R;D3] ) [NH;!R;D2](-[C;!R;D3])-[CH;!R;D3] [C;!R;D3](-[NH;!R;D2])(=[O;!R;D1])-[O;!R;D2] ( = [O;!R;D1]=[C;!R;D3] ) [O;!R;D2](-[C;!R;D3])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[O;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2]  [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] ( = ( ( ( = ) ( [Br;!R;D1]-[c;R;D3] ) ( = [O;!R;D1]=[C;!R;D3] ) ( = ) ( = [O;!R;D1]=[C;!R;D3] ) ( 2 ) ( [F;!R;D1]-[C;!R;D4] ) ( = ) ( = ) ( = [O;!R;D1]=[C;!R;D3] ) ( = [O;!R;D1]=[C;!R;D3] ) ( [F;!R;D1]-[C;!R;D4] ) ( = ( = ( = [O;!R;D1]=[C;!R;D3] ) [c;R;D3](-[F;!R;D1])(:[c;R;D3]):[c;R;D3] 1 ) [Cl;!R;D1]-[C;!R;D4] ) ( = [O;!R;D1]=[C;!R;D3] ) [c;R;D3](-[F;!R;D1])(:[c;R;D3]):[c;R;D3] 1 ) [c;R;D3](-[F;!R;D1])(:[c;R;D3]):[c;R;D3] 1 ) [c;R;D3](-[F;!R;D1])(:[c;R;D3]):[c;R;D3] 1 ) [c;R;D3](-[F;!R;D1])(:[c;R;D3]):[c;R;D3] 1 [CH2;R;D2](-[CH2;R;D2])-[CH;R;D3] [CH2;R;D2](-[CH2;R;D2])-[CH2;R;D2] [CH2;R;D2](-[CH2;R;D2])-[CH2;R;D2] [CH2;R;D2](-[CH2;R;D2])-[CH2;R;D2] [CH2;R;D2](-[CH2;R;D2])-[CH2;R;D2] [CH2;R;D2](-[CH2;R;D2])-[CH2;R;D2] [CH2;R;D2](-[CH2;R;D2])-[CH2;R;D2]  [CH2;R;D2](-[CH2;R;D2])-[CH2;R;D2] [CH2;R;D2](-[CH2;R;D2])-[CH2;R;D2] [CH2;R;D2](-[CH2;R;D2])-[CH2;R;D2] [CH2;R;D2](-[CH2;R;D2])-[CH2;R;D2] [CH2;R;D2](-[CH2;R;D2])-[CH2;R;D2] [CH2;R;D2](-[CH2;R;D2])-[CH2;R;D2]"

problem_smarts = "[CH3;!R;D1]-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH3;!R;D1] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[CH2;!R;D2] [CH2;!R;D2](-[CH2;!R;D2])-[O;!R;D2] [O;!R;D2](-[C;!R;D3])-[CH2;!R;D2] [C;!R;D3](=[O;!R;D1])(-[O;!R;D2])-[c;R;D3] ( = [O;!R;D1]=[C;!R;D3] ) [c;R;D3](-[C;!R;D3])(:[c;R;D3]):[cH;R;D2] 1 [cH;R;D2](:[c;R;D3]):[cH;R;D2] [cH;R;D2](:[cH;R;D2]):[cH;R;D2] [cH;R;D2](:[cH;R;D2]):[cH;R;D2] [cH;R;D2](:[c;R;D3]):[cH;R;D2] 1 ) [c;R;D3](-[C;!R;D3])(:[c;R;D3]):[cH;R;D2] 1 ) [c;R;D3](-[F;!R;D1])(:[c;R;D3]):[c;R;D3] ( = [O;!R;D1]=[C;!R;D3]"


def convert2smiles(new_tokens_list):
    pattern =  "(\[[^\]]+]|Br?|Cl?|N|O|S|P|F|I|b|c|n|o|s|p|\(|\)|\.|=|#|-|\+|\\\\|\/|:|~|@|\?|>|\*|\$|\%[0-9]{2}|[0-9])"
    regex = re.compile(pattern)
    smarts = ""
    for new_token in new_tokens_list.split():
        symbols = [token for token in regex.findall(new_token)]
        smarts += symbols[0]

   #RDLogger.DisableLog('rdApp.*')
    try:
        mol_sma = Chem.MolFromSmarts(smarts) #=> the error comes with this line
        if mol_sma is not None:
            return Chem.MolToSmiles(mol_sma)
        else:
            return None
    except:
        return None
    

if __name__=="__main__":
    for n, smarts in enumerate([normal_smarts, still_ok_smarts, problem_smarts]):
        smiles = convert2smiles(smarts)
        if smiles:
            print(f"{n} {smiles = }")
        else:
            print(f"{n} ERROR:\t{smiles = }")
 ```

            
**Expected behavior**
In the above code snippet, `normal_smarts` converted to SMILES without any error and even though `still_ok_smarts` can't be converted to SMILES, the RDKit Error can be avoided with ```try; exception``` or ```RDLogger.DisableLog('rdApp.*')```
`problem_smarts` cause an error that I was unable to avoid. How can the error be avoided? 

**Screenshots**
```
*** Error in `python': double free or corruption (out): 0x00007fffe28774e0 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x81299)[0x7fbf96bfb299]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/Chem/../../../../libRDKitSmilesParse.so.1(+0x28a16)[0x7fbf7addba16]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/Chem/../../../../libRDKitSmilesParse.so.1(_ZN5RDKit11SmartsToMolERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEibPSt3mapIS5_S5_St4lessIS5_ESaISt4pairIS6_S5_EEE+0x1c8)[0x7fbf7ade1648]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/Chem/rdmolfiles.so(_ZN5RDKit13MolFromSmartsEN5boost6python3api6objectEbNS1_4dictE+0x5b6)[0x7fbf7ab12186]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/Chem/rdmolfiles.so(+0x74c82)[0x7fbf7aae8c82]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/../../../libboost_python38.so.1.74.0(_ZNK5boost6python7objects8function4callEP7_objectS4_+0x2ed)[0x7fbf8fb66eed]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/../../../libboost_python38.so.1.74.0(+0x210f9)[0x7fbf8fb670f9]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/../../../libboost_python38.so.1.74.0(_ZNK5boost6python6detail17exception_handlerclERKNS_9function0IvEE+0x63)[0x7fbf8fb6bef3]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/Chem/rdmolfiles.so(+0x6bc49)[0x7fbf7aadfc49]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/../../../libboost_python38.so.1.74.0(_ZNK5boost6python6detail17exception_handlerclERKNS_9function0IvEE+0x37)[0x7fbf8fb6bec7]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/Chem/rdmolfiles.so(+0x6bbe9)[0x7fbf7aadfbe9]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/../../../libboost_python38.so.1.74.0(_ZNK5boost6python6detail17exception_handlerclERKNS_9function0IvEE+0x37)[0x7fbf8fb6bec7]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/Chem/rdchem.so(+0x1654a9)[0x7fbf7afd74a9]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/../../../libboost_python38.so.1.74.0(_ZNK5boost6python6detail17exception_handlerclERKNS_9function0IvEE+0x37)[0x7fbf8fb6bec7]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/Chem/rdchem.so(+0x10a87d)[0x7fbf7af7c87d]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/../../../libboost_python38.so.1.74.0(_ZNK5boost6python6detail17exception_handlerclERKNS_9function0IvEE+0x37)[0x7fbf8fb6bec7]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/Chem/rdchem.so(+0x10a4bd)[0x7fbf7af7c4bd]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/../../../libboost_python38.so.1.74.0(_ZNK5boost6python6detail17exception_handlerclERKNS_9function0IvEE+0x37)[0x7fbf8fb6bec7]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/Chem/rdchem.so(+0x10a0fd)[0x7fbf7af7c0fd]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/../../../libboost_python38.so.1.74.0(_ZNK5boost6python6detail17exception_handlerclERKNS_9function0IvEE+0x37)[0x7fbf8fb6bec7]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/Chem/rdchem.so(+0x109d3d)[0x7fbf7af7bd3d]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/../../../libboost_python38.so.1.74.0(_ZNK5boost6python6detail17exception_handlerclERKNS_9function0IvEE+0x37)[0x7fbf8fb6bec7]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/Chem/rdchem.so(+0x10997d)[0x7fbf7af7b97d]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/../../../libboost_python38.so.1.74.0(_ZNK5boost6python6detail17exception_handlerclERKNS_9function0IvEE+0x37)[0x7fbf8fb6bec7]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/rdBase.so(+0x5d419)[0x7fbf8fbe2419]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/../../../libboost_python38.so.1.74.0(_ZNK5boost6python6detail17exception_handlerclERKNS_9function0IvEE+0x37)[0x7fbf8fb6bec7]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/rdBase.so(+0x5d3b9)[0x7fbf8fbe23b9]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/../../../libboost_python38.so.1.74.0(_ZNK5boost6python6detail17exception_handlerclERKNS_9function0IvEE+0x37)[0x7fbf8fb6bec7]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/rdBase.so(+0x5d359)[0x7fbf8fbe2359]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/../../../libboost_python38.so.1.74.0(_ZNK5boost6python6detail17exception_handlerclERKNS_9function0IvEE+0x37)[0x7fbf8fb6bec7]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/rdBase.so(+0x5d2f9)[0x7fbf8fbe22f9]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/../../../libboost_python38.so.1.74.0(_ZN5boost6python21handle_exception_implENS_9function0IvEE+0x3f)[0x7fbf8fb6be0f]
/home/azkiradam/anaconda3/envs/pytorch/lib/python3.8/site-packages/rdkit/../../../libboost_python38.so.1.74.0(+0x1e093)[0x7fbf8fb64093]
python(_PyObject_MakeTpCall+0x23c)[0x55e1bb8e6dbc]
python(_PyEval_EvalFrameDefault+0x4596)[0x55e1bb972666]
python(_PyFunction_Vectorcall+0x10b)[0x55e1bb93cfcb]
python(+0x1035db)[0x55e1bb8a65db]
python(_PyEval_EvalCodeWithName+0x300)[0x55e1bb93c270]
python(PyEval_EvalCode+0x23)[0x55e1bb9d1543]
python(+0x22e5e4)[0x55e1bb9d15e4]
python(+0x254854)[0x55e1bb9f7854]
python(+0x115390)[0x55e1bb8b8390]
python(PyRun_SimpleFileExFlags+0x384)[0x55e1bb8bb0d2]
python(+0x118bf0)[0x55e1bb8bbbf0]
python(Py_BytesMain+0x39)[0x55e1bb9faa09]
/lib64/libc.so.6(__libc_start_main+0xf5)[0x7fbf96b9c555]
python(+0x1e6fe5)[0x55e1bb989fe5]

```

**Configuration (please complete the following information):**
 - RDKit version: **2021.03.1**
 - OS: **CentOS Linux 7** 
 - Python version (if relevant): **python3.8**
 - Are you using conda?
 - If you are using conda, which channel did you install the rdkit from?
`conda install -c conda-forge rdkit`

**Additional context**

**Comment by bp-kelley** [comment]
I can duplicate the error on ubuntu.  This error is non-recoverable unfortunately.

**Comment by bp-kelley** [comment]
I've isolated a simpler condition:

```
>>> from rdkit import Chem
>>> Chem.MolFromSmarts("c;R;D3]1[cH;R;D2][cH;R;D2]1)[c;R;D3]1)")
```

It looks like a ring closure followed by a unmatched closing branch is causing the issue.

**Comment by bp-kelley** [comment]
Correction, two unmatched closing branches:

```
>>> from rdkit import Chem
>>> Chem.MolFromSmarts("c1cc1))")
```

**Comment by greglandrum** [comment]
FWIW, I cannot reproduce what Brian is observing with the release version of RDKit on linux

**Comment by bp-kelley** [comment]
@ashyrmamatov01 @greglandrum This bug exists in 2021.03.1 but has been fixed in 2021.03.3 so a simple upgrade should solve your issue.

**Comment by greglandrum** [comment]
I'm closing this because it looks like the question is answered. Please reopen if necessary.
